### PR TITLE
J22: Agrus Kos, Eternal Soldier and support

### PR DESCRIPTION
**Pull request by Northmoc**

### CODE REVIEW

File: forge-game/src/main/java/forge/game/ability/effects/CopySpellAbilityEffect.java [hunk 1]
```diff
@@ -97,7 +97,8 @@ public void resolve(SpellAbility sa) {
                 // Find subability or rootability that has targets
                 SpellAbility targetedSA = chosenSA;
                 while (targetedSA != null) {
-                    if (targetedSA.usesTargeting() && targetedSA.getTargets().size() != 0) {
+                    if (targetedSA.usesTargeting() && (targetedSA.getTargets().size() != 0
+                            || (sa.hasParam("Defined") && sa.getParam("Defined").startsWith("Triggered")))) {
```
Comment (Northmoc) [inline forge-game/src/main/java/forge/game/ability/effects/CopySpellAbilityEffect.java:101]: when Forge references targetedSA for Agrus Kos, targetSA.getTargets().size() always returns 0 - hence this hack

not sure this is the best way to solve this though for this card, as BecomesTarget really has done all the checks we need on whether the right thing is targeted

Comment (Hanmac) [reply forge-game/src/main/java/forge/game/ability/effects/CopySpellAbilityEffect.java:101]: What ability did you test? Triggered or Activated?

Comment (Northmoc) [reply forge-game/src/main/java/forge/game/ability/effects/CopySpellAbilityEffect.java:101]: Main test was activated ability of Kabuto Moth

Comment (Hanmac) [reply forge-game/src/main/java/forge/game/ability/effects/CopySpellAbilityEffect.java:101]: You should debug why that is the case, does it have something to do with becomes target trigger?

Comment (tool4ever) [reply forge-game/src/main/java/forge/game/ability/effects/CopySpellAbilityEffect.java:101]: The problem is multiple (from same source) `AbilityActivated` on the stack are usually the same object

What you can try:

Add `AbilityKey.StackInstance` like `TriggerSpellAbilityCastOrCopy` already does
\+ `Defined$ TriggeredStackInstance`

Comment (Hanmac) [reply forge-game/src/main/java/forge/game/ability/effects/CopySpellAbilityEffect.java:101]: if that's the case we really need to cleanup the Core more about this

Comment (Northmoc) [reply forge-game/src/main/java/forge/game/ability/effects/CopySpellAbilityEffect.java]: @tool4ever your suggestion seems fine

I think remaining core issues could go to an issue

**Comment by Agetian** [comment]
:+1:

File: forge-game/src/main/java/forge/game/trigger/TriggerBecomesTarget.java [hunk 2]
```diff
@@ -85,6 +85,7 @@ public final boolean performTest(final Map<AbilityKey, Object> runParams) {
     public final void setTriggeringObjects(final SpellAbility sa, Map<AbilityKey, Object> runParams) {
         sa.setTriggeringObject(AbilityKey.Source, ((SpellAbility) runParams.get(AbilityKey.SourceSA)).getHostCard());
         sa.setTriggeringObjectsFrom(runParams, AbilityKey.SourceSA, AbilityKey.Target);
+        sa.setTriggeringObject(AbilityKey.StackInstance, sa.getHostCard().getGame().getStack().getInstanceFromSpellAbility((SpellAbility) runParams.get(AbilityKey.SourceSA)));
```
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/trigger/TriggerBecomesTarget.java:88]: i hope that getInstanceFromSpellAbility doesn't make us more problems, but i seems to work for now

**Pull request merged**